fenv_lmi.cpp

```diff
@@ -190,44 +175,82 @@ void fenv_rounding(e_ieee754_rounding rounding_mode)
     intel_control_word control_word(x87_control_word());
     control_word.rc(rc);
     x87_control_word(control_word.cw());
-#else  // Unknown compiler or platform.
-#   error Unknown compiler or platform.
-#endif // Unknown compiler or platform.
-}
-
-bool fenv_is_valid()
-{
-#if defined LMI_X86
-    return default_x87_control_word() == x87_control_word();
-#else  // Unknown compiler or platform.
-#   error Unknown compiler or platform.
-#endif // Unknown compiler or platform.
+#endif // !defined __BORLANDC__
+#else  // !defined LMI_X87
+    int z =
+          (fe_tonearest  == rounding_mode) ? FE_TONEAREST
+        : (fe_downward   == rounding_mode) ? FE_DOWNWARD
+        : (fe_upward     == rounding_mode) ? FE_UPWARD
+        : (fe_towardzero == rounding_mode) ? FE_TOWARDZERO
+        : throw std::runtime_error("Failed to set rounding mode.")
+        ;
+    fesetround(z);
+#endif // !defined LMI_X87
 }
 
 namespace
 {
-std::string fenv_explain_invalid_control_word()
+bool check_fpenv_and_explain_if_invalid(std::string& explanation)
 {
     std::ostringstream oss;
-    oss
-        << "The floating-point control word was unexpectedly '"
-        << std::hex << std::internal << std::showbase << std::setfill('0')
-#if defined LMI_X86
-        << std::setw(6) << x87_control_word()
-#else  // Unknown compiler or platform.
-#   error Unknown compiler or platform.
-#endif // Unknown compiler or platform.
-        << "'."
-        << "\nProbably some other program changed this crucial setting."
-        << "\nIt has been reset correctly. Rerun any illustration that"
-        << "\nwas being run when this message appeared, because it may"
-        << "\nbe incorrect."
-        << "\n"
-        ;
-    return oss.str();
+
+#if defined LMI_X87
+    if(default_x87_control_word() != x87_control_word())
+        {
+        oss
+            << "The floating-point control word was unexpectedly '"
+            << std::hex << std::internal << std::showbase << std::setfill('0')
+            << std::setw(6) << x87_control_word()
+            << "'.\n"
+            ;
+        }
+#else  // !defined LMI_X87
+    if(fenv_rounding() != fe_tonearest)
+        {
+        oss
+            << "The rounding mode was unexpectedly "
+            << fenv_rounding()
+            << " instead of the expected "
+            << fe_tonearest
+            << ".\n"
+            ;
+        }
+
+    int const enabled_fp_exceptions = fetestexcept(FE_ALL_EXCEPT);
+    if(enabled_fp_exceptions != 0)
```

Comment:
I hesitated for quite some before adding this code because I don't think this changes the results of any computations. But OTOH it could do it if some library function checks for FP exceptions and it seems to be correspond better to the spirit of the existing x87 code.